### PR TITLE
socket: Enforce socket length

### LINKED ISSUE #268: enforce socket length
Only some parts of the virtcontainers code enforce unix domain socket length.

### PR DESCRIPTION
**Pull request by jodh-intel**
A Unix domain socket is limited to 107 usable bytes on Linux. However,
not all code creating socket paths was checking for this limits.

Created a new `buildSocketPath()` function (with tests) to encapsulate
the logic and updated all code creating sockets to use it.

Fixes #268.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
This PR raises a few issues:

- We're already almost at the `107`-byte path limit.
  - #200 made the problem worse by increasing the `storagePathSuffix` path by 5 bytes (`len("sandboxes")-len("pods")`)
  - We should consider changing `storagePathSuffix` from `"/virtcontainers/sandboxes"` to `"/kata/pods"`, saving `15` bytes.
- The changes to `qmpSocketPath()` are frankly gross but I've attempted to re-implement what was there before. The logic is indicative of the lack of remaining bytes available for the files we need to create.
- We should have found this bug earlier (aka we need more unit tests in virtcontainers).

Related: #257.

File: virtcontainers/utils.go [hunk 1]
```diff
@@ -7,15 +7,22 @@ package virtcontainers
...
 )
 
 const cpBinaryName = "cp"
 
 const fileMode0755 = os.FileMode(0755)
 
+// Effective maximum Unix domain socket length
+//
+// See unix(7).
+const maxSocketPathLen = 107
```
Comment (grahamwhaley) [inline virtcontainers/utils.go]: I wish there was some go system level const for this eh?
Found a maybe relevant ticket https://github.com/golang/go/issues/6895 - and, looks like @jessfraz was around this area back in 2015 maybe - so maybe can give some insight/input.

**Comment by jodh-intel** [comment]
The tests are failing, but atleast the code is working! :smile:

```
--- FAIL: TestCreateSandboxKataAgentSuccessful (0.00s)
	api_test.go:334: path too long (got 109, max 107): /tmp/virtcontainers-tmp-708439037/virtcontainers/sandboxes/run/7f49d00d-1995-4156-8c79-5f5ab24ce138/kata.sock

```

**Comment by grahamwhaley** [comment]
Heh heh @jodh-intel 
So, looking at that path, the most obvious fix that springs to my mind is s/virtcontainers/vc/ - which would claim us back 24 chars or so I reckon...

**Comment by grahamwhaley** [comment]
@jodh-intel - if the tests are failing due to the new check, does that mean we cannot merge this until we have fixed some other code (to reduce the path length)? If so, which I think, then I'm going to mark this as DNM to make it clear we are waiting for something else. Please do revert that if I've mis-understood...

**Comment by sboeuf** [comment]
Leaving the decision to @jodh-intel to merge it, but completely fine with me ;)

**Comment by WeiZhang555** [comment]
Codes LGTM. But need rebase :-)

**Comment by jodh-intel** [comment]
Branch rebased. I've raised #290 to reduce the path lengths.

**Pull request merged**